c/shared/source/uforead/uforead.c

```diff
@@ -401,12 +401,61 @@ void ufoFree(ufoCtx h) {
     dnaFREE(h->chars.byName);
     dnaFREE(h->chars.widths);
     dnaFREE(h->tmp);
+
+    for (int i = 0; i < h->data.glifRecs.cnt; i++){
+        GLIF_Rec rec = h->data.glifRecs.array[i];
+        /* rec.glyphName is not allocated on its own, so no need to free. */
+        if (rec.glifFileName != NULL)
+            memFree(h, rec.glifFileName);
+        if (rec.glifFilePath != NULL)
+            memFree(h, rec.glifFilePath);
+        if (rec.altLayerGlifFileName != NULL)
+            memFree(h, rec.altLayerGlifFileName);
+    }
     dnaFREE(h->data.glifRecs);
+
+    for (int i = 0; i < h->data.glifOrder.cnt; i++){
+        GlIFOrderRec rec = h->data.glifOrder.array[i];
+        if (rec.glyphName != NULL)
+            memFree(h, rec.glyphName);
+    }
     dnaFREE(h->data.glifOrder);
+
+    for (int i = 0; i < h->data.opList.cnt; i++){
+        OpRec rec = h->data.opList.array[i];
+        if (rec.pointName != NULL)
+            memFree(h, rec.pointName);
+    }
     dnaFREE(h->data.opList);
     freeStrings(h);
     dnaFree(h->dna);
 
+    if (h->top.cid.CIDFontName.ptr != NULL)
```

Comment:
I haven't gone too far in depth with the analysis but I'm worried about these lines and the shared responsibility over the `top` structure. 

First and most basically it seems like if this cleanup is going to happen at all it should be in a function in `shared/source/absfont/absfont.c`, probably called `abfFreeTopDict()` to match `abfInitTopDict()`. 

But that's a detail. The more pressing problem is that `top` is typically passed into `ufoBegFont()` as a double-pointer, which is then initialized and also stored in the `ufoCtx`. This is not ideal to begin with -- it would be preferable if `ufoBegFont()` could set the UFO `h->top` back to NULL before returning -- but if the ctx will still be used and needs it that may be impractical.

Let's assume that the double-pointer `top` is only valid a) while the ufoCtx is allocated and b) between calls to `ufoBegFont()` (It seems like one can call `ufoEndFont()` and not loose anything, as that's just for closing files and such?) 

In that case the ufoCtx top structure may be re-used (right? there's only one in the `txCtx` ...), and that's a problem given that `abfInitFontDict()` does't check for and free already-allocated values. A normal thing here would be to have `abfReInitTopDict()` which calls `abfFreeTopDict()` first and then `abfInitTopDict()`, and just use the `ReInit` version when you may be in a reuse situation. But the code could also just call `abfFreeTopDict()` first. 

Doing this safely is going to require a bit of analysis of the specific lifetime of the `top` structure.

All of these comments also apply to `fdict`.